include/alarm_heap.hpp

```diff
@@ -0,0 +1,198 @@
+  }
+
+  const unsigned int index() { return _index; }
+  AlarmDef::Severity severity() { return _severity; }
+  void set_severity(AlarmDef::Severity severity) { _severity = severity; }
+  bool in_heap() { return (_heap != nullptr); }
+  uint64_t get_pop_time() const { return _pop_time; }
+
+private:
+  uint64_t _pop_time;
+  AlarmDef::Severity _severity;
+  unsigned int _index;
+};
+
+class AlarmInfo
+{
+public:
+  AlarmInfo(AlarmInHeap* alarm_in_heap,
+            AlarmDef::Severity severity) :
+    _alarm_in_heap(alarm_in_heap),
+    _severity(severity)
+  {}
+
+  ~AlarmInfo();
+
+  // Compares a new severity with the current severity for this alarm,
+  // and returns the severity change
+  // @params severity - the new severity
+  // @returns         - the severity change
+  AlarmSeverityChange new_alarm_severity(AlarmDef::Severity severity);
+
+  // Updates the alarm in the heap's severity, and its time to pop
+  void update_alarm_in_heap(AlarmDef::Severity new_severity,
+                            uint64_t time_to_delay);
+
+  AlarmInHeap* alarm_in_heap() { return _alarm_in_heap; }
+  AlarmDef::Severity get_severity() { return _severity; }
+  void set_severity(AlarmTableDef& alarm_table_def);
+
+  // Determines whether this alarm should be resent after failure
+  bool should_resend_alarm(AlarmDef::Severity severity)
+  {
+    // Resend the alarm if its severity and the current severity match,
+    // and there isn't already an alarm in the heap waiting to update
+    // the severity.
+    return ((_severity == severity) && (!_alarm_in_heap->in_heap()));
+  }
+
+private:
+  AlarmInHeap* _alarm_in_heap;
+  AlarmDef::Severity _severity;
+};
+
+typedef std::map<unsigned int, AlarmInfo*> AllAlarmsStateMap;
+
+// This class accepts alarm triggers, checks that they correspond to valid
+// alarms, and passes valid alarms to the alarm trap sender to send as SNMP
+// INFORMs. It also filters the alarms.
+class AlarmHeap
+{
+public:
+  // Times to send an alarms. If we're reducing the severity of an alarm we
+  // add a delay of 30secs (to protect against flicker). Increased severity
+  // alarms or resyncing alarms are sent immediately.
+  const static uint64_t ALARM_REDUCED = 30000;
+  const static uint64_t ALARM_RETRY_DELAY = 30000;
+  const static uint64_t ALARM_RESYNC_DELAY = 0;
+  const static uint64_t ALARM_INCREASED = 0;
+
+  // Constructor/Destructor
+  AlarmHeap(AlarmTableDefs* alarm_table_defs);
+  virtual ~AlarmHeap();
+
+  // Generates an alarmActiveState inform if the identified alarm is not
+  // of CLEARED severity and not already active (or subject to filtering).
+  // Generates an alarmClearState inform if the identified alarm is of a
+  // CLEARED severity and an associated alarm is active (unless subject
+  // to filtering).
+  virtual void issue_alarm(const std::string& issuer,
+                   const std::string& identifier);
+
+  static void* heap_sender_function(void* data);
+
+  void heap_sender();
```

Comment:
I was going to complain that this wasn't a very good name, but the best i could come up with was `popping_loop`.